lib/redbird/plug/session/redis.ex

```diff
@@ -25,17 +24,24 @@ defmodule Plug.Session.REDIS do
   end
 
   def put(conn, nil, data, init_options) do
-    put(conn, Key.generate(), data, init_options)
+    key =
+      Key.generate()
+      |> Key.sign_key(conn)
+
+    put(conn, key, data, init_options)
   end
 
   def put(conn, key, data, init_options) do
-    key
-    |> Key.sign_key(conn)
-    |> set_key_with_retries(Value.serialize(data), session_expiration(init_options), 1)
+    if Key.accessible?(key, conn) do
```

Comment:
Do you know if we need to make this check here? 

Looking at [Plug.Session.Store's docs](https://hexdocs.pm/plug/Plug.Session.Store.html#c:put/4), I see that we're supposed to generate the key when its value is `nil` (which is what you're doing in the function above). But I think after that we can just take the `key` that Plug is giving us and setting it.

At the very least, I don't think we should be returning the `key` in the `else` clause below since it hasn't passed the `Key.accessible?` check. In that case, we're not actually setting the key, but by returning it, we're incorrectly telling Plug that we've storing the session. 

So, do you know if there's harm in just having the `put` function have these two function heads?

```elixir
  def put(conn, nil, data, init_options)
    key = Key.generate() |> Key.sign_key(conn)

    put(conn, key, data, init_options)
  end

  def put(_conn, key, data, init_options) do
    set_key_with_retries(key, Value.serialize(data), session_expiration(init_options), 1)
  end
```

Reply:
The `key` that Plug gives us is user-controlled—it's whatever the user sent in the cookie (or `nil`, if the user didn't send the cookie.) This of course is why the signature needs to be verified as part of `get`—but I'm pretty sure the same reasoning applies to `put`. If we don't verify the signature, it seems to me that it would allow the user to set arbitrary keys in redis, which would be a pretty bad security flaw in its own right.

> At the very least, I don't think we should be returning the key in the else clause below since it hasn't passed the Key.accessible? check. In that case, we're not actually setting the key, but by returning it, we're incorrectly telling Plug that we've storing the session.

We have only two options here: we can either return some key, or we can throw an exception. `plug` [puts](https://github.com/elixir-plug/plug/blob/39e1cede12a55496af892bac0c0d7027f5be87ba/lib/plug/session.ex#L96-L97) the value we return directly into the cookie—we don't exactly have a clean way to signal an error.

But we could raise an exception, if you think that's better. Failing silently on a non-valid key is what we do in [`delete/3`](https://github.com/importfeed/redbird/blob/5cabf061d1911461f4c0598dcd3d78be1fc39465/lib/redbird/plug/session/redis.ex#L43-L47)—it would make sense to be consistent about it.

(Actually, for full consistency, if we want an invalid key to throw an error on `put/4` and `delete/3`, wouldn't it make sense to throw an error also on `get/3`, instead of just returning `nil` as if the key was valid but the value was unset?)

Come to think of it, another option would be to treat an invalid (badly-signed, or unsigned) key exactly like how we treat `nil`, and generate a new key to use. That's probably the best solution here.

Reply:
@foxbenjaminfox thanks so much for all that information. 

>  I'm pretty sure the same reasoning applies to put. If we don't verify the signature, it seems to me that it would allow the user to set arbitrary keys in redis, which would be a pretty bad security flaw in its own right.

That makes sense. So let's keep the verification step in `put` just like you have it now. 

> Come to think of it, another option would be to treat an invalid (badly-signed, or unsigned) key exactly like how we treat nil, and generate a new key to use. That's probably the best solution here.

I like this idea. I don't know if that's somewhat unexpected from the end-user's perspective, but on the other hand, the end user should only care that they get a key back where we've stored the session. So I think that idea sounds great. 

So let's go ahead and treat an invalid key exactly like we treat `nil`. Once that change is in place, I think this is good to go.

Reply:
> So let's go ahead and treat an invalid key exactly like we treat nil. Once that change is in place, I think this is good to go.

6cc04fd75f270f does this by having `put/4` call itself with `nil` as the second argument if it can't validate the key.